plugins/woocommerce-admin/client/products/fields/category-field/category-field.tsx

```diff
@@ -0,0 +1,249 @@
+
+	const onInputChange = ( searchString?: string ) => {
+		setSearchValue( searchString || '' );
+		searchCategories( searchString || '' );
+	};
+
+	const searchDelayed = useMemo(
+		() => debounce( onInputChange, 150 ),
+		[ onInputChange ]
+	);
+
+	const onSelect = ( item: ProductCategory, selected: boolean ) => {
+		if ( selected ) {
+			onChange(
+				getSelectedWithParents(
+					[ ...value ],
+					item,
+					categoryTreeKeyValues
+				)
+			);
+		} else {
+			onChange( value.filter( ( i ) => i.id !== item.id ) );
+		}
+	};
+
+	const selectedIds = value.map( ( item ) => item.id );
+	let selectControlItems = categoriesSelectList;
+	// Add the add new button if search value does not exist.
+	if (
+		searchValue.length > 0 &&
+		categoriesSelectList.length > 0 &&
+		! categoriesSelectList.find(
+			( cat ) =>
+				cat.label.toLowerCase() === searchValue.toLowerCase() &&
+				cat.value !== 'add-new'
+		)
+	) {
+		if (
+			! categoriesSelectList.find( ( cat ) => cat.value === 'add-new' )
+		) {
+			selectControlItems.push( { value: 'add-new', label: searchValue } );
+		}
+	} else {
+		selectControlItems = categoriesSelectList.filter(
+			( cat ) => cat.value !== 'add-new'
+		);
+	}
+	// converted selected items to the select control format.
+	const selected = ( value || [] ).map( ( cat ) => ( {
+		value: cat.id.toString(),
+		label: cat.name,
+	} ) );
+
+	return (
+		<SelectControl
+			multiple
+			items={ selectControlItems }
+			label={ label }
+			selected={ selected }
+			onSelect={ ( item: SelectControlItem ) =>
+				item &&
+				onSelect(
+					categoryTreeKeyValues[ parseInt( item.value, 10 ) ].data,
+					! selectedIds.includes( parseInt( item.value, 10 ) )
+				)
+			}
+			onRemove={ ( item: SelectControlItem ) =>
+				item &&
+				onSelect(
+					categoryTreeKeyValues[ parseInt( item.value, 10 ) ].data,
+					false
+				)
+			}
+			onInputChange={ searchDelayed }
+			getFilteredItems={ getFilteredItems }
+			placeholder={ selected.length === 0 ? placeholder : '' }
+			keepMenuOpenOnSelect={ true }
+		>
+			{ ( {
+				items,
+				isOpen,
+				getMenuProps,
+				getToggleButtonProps,
+				selectItem,
+				highlightedIndex,
+				setInputValue,
+			} ) => {
+				return (
+					<>
+						<div
+							{ ...getMenuProps() }
+							className={ classnames(
+								'woocommerce-select-control__menu',
+								{
+									'is-open': isOpen,
+									'category-field-dropdown__menu': isOpen,
+								}
+							) }
+						>
+							{ isOpen && isSearching && items.length === 0 && (
+								<div className="category-field-dropdown__item">
+									<div className="category-field-dropdown__item-content">
+										<Spinner />
+									</div>
+								</div>
+							) }
+							{ isOpen &&
+								items.length > 0 &&
+								items
+									.filter(
+										( item ) =>
+											categoryTreeKeyValues[
+												parseInt( item.value, 10 )
+											]?.parentID === 0 ||
+											item.value === 'add-new'
```

Comment:
In order for the selection to work well and the keyboard selection. We had to pass in the entire list of categories to the component, this includes all the children as well.
**But**
We don't want to display all of them in the top list, so we filter out all the children here.

You can see that for the `CategoryFieldItem` we actually pass in:
```
categoryTreeKeyValues[
  parseInt(
    item.value,
    10
  )
]
```
This is for the same reason as above, this actually returns us a tree like object that contains children.